dc/s2n-quic-dc/src/stream/server/tokio/tcp/worker.rs

```diff
@@ -538,3 +567,274 @@ impl WorkerState {
+                        }
+                    },
+                    Poll::Pending => {
+                        event_data.blocked_count += 1;
+                        return Poll::Pending;
+                    }
+                },
+            };
+
+            // try to read an initial packet from the socket
+            let res = {
+                let (stream, remote_address) = stream.as_mut().unwrap();
+                WorkerState::poll_initial_packet(
+                    cx,
+                    stream,
+                    remote_address,
+                    recv_buffer,
+                    sojourn_time,
+                    publisher,
+                )
+            };
+
+            let Poll::Ready(res) = res else {
+                // if we got `Pending` but we don't own the recv buffer then we need to copy it
+                // into the worker so we can resume where we left off last time
+                if blocked_count == 0 {
+                    let buffer = recv_buffer.take();
+                    *state = WorkerState::Buffering {
+                        buffer,
+                        blocked_count,
+                    };
+                }
+
+                if let WorkerState::Buffering { blocked_count, .. } = state {
+                    *blocked_count += 1;
+                }
+
+                return Poll::Pending;
+            };
+
+            let initial_packet = res?;
+
+            let (socket, remote_address) = stream.take().unwrap();
+
+            let recv_buffer = recv_buffer.make_contiguous();
+
+            let mut secret_control = vec![];
+            let credentials = &initial_packet.credentials;
+
+            let map = &context.secrets;
+            let Some((export_secret, ciphersuite, keys, application_params)) = map
+                .secret_for_credentials(
+                    credentials,
+                    initial_packet.source_queue_id,
+                    &TransportFeatures::TCP,
+                    &mut secret_control,
+                )
+            else {
+                let error = io::Error::new(
+                    io::ErrorKind::NotFound,
+                    format!("missing credentials for client: {credentials:?}"),
+                );
+
+                if !secret_control.is_empty() {
+                    *stream = Some((socket, remote_address));
+                    *state = WorkerState::Erroring {
+                        offset: 0,
+                        buffer: secret_control,
+                        error,
+                    };
+                    continue;
+                } else {
+                    // Close socket immediately
+                    let _ = socket.set_linger(Some(Duration::ZERO));
+                    drop(socket);
+                }
+                return Err(Some(error)).into();
+            };
+
+            if keys
+                .application
+                .opener
+                .on_decrypt_success(recv_buffer.into())
```

Comment:
For a future PR: we will probably want to replace this with a call to decrypt(...), since we ought to verify that this is actually a valid packet before we mark it as used in the replay protection map.